File: sdk/cpp/gnmi/tests/test_gnmi_service.cpp
/// YANG Development Kit
// Copyright 2016 Cisco Systems. All rights reserved
//
////////////////////////////////////////////////////////////////
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
//////////////////////////////////////////////////////////////////
#include <thread>
#include <chrono>

#include <ydk/gnmi_provider.hpp>
#include <ydk/gnmi_service.hpp>
#include <ydk/codec_provider.hpp>
#include <ydk/codec_service.hpp>
#include <ydk/entity_data_node_walker.hpp>

#include <ydk/filters.hpp>

#include "test_utils.hpp"

#include "../../core/src/catch.hpp"
#include "../../core/tests/config.hpp"

using namespace std;
using namespace ydk;
using namespace path;
using namespace ydktest;

static string cap_update = R"(
      {
        "name": "openconfig-bgp",
        "organization": "OpenConfig working group",
        "version": "2016-06-21"
      },)";

TEST_CASE("gnmi_service_capabilities")
{
    // session
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;

    gNMIServiceProvider provider{repo, address, port, "admin", "admin"};
    gNMIService gs{};

    string json_caps = gs.capabilities(provider);
    //std::cout << json_caps << std::endl;
    REQUIRE(json_caps.find(cap_update) != string::npos);
}

static string int_update = R"([{\"name\":\"Loopback10\",\"config\":{\"name\":\"Loopback10\",\"description\":\"Test\"}}])";

static string bgp_update = R"([{\"neighbor-address\":\"172.16.255.2\",\"config\":{\"neighbor-address\":\"172.16.255.2\",\"peer-as\":65172}}])";

void read_sub(const char * subscribe_response)
{
    cout << endl << subscribe_response << endl;
}

void gnmi_service_subscribe_callback(const char * subscribe_response)
{
	//read_sub(subscribe_response);
	string response = subscribe_response;
	REQUIRE(response.find(int_update) != string::npos);
}

void gnmi_service_subscribe_multiples_callback(const char * subscribe_response)
{
	//read_sub(subscribe_response);
	string response = subscribe_response;
	REQUIRE(response.find(int_update) != string::npos);
    REQUIRE(response.find(bgp_update) != string::npos);
}

TEST_CASE("gnmi_service_subscribe")
{
    // session
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;

    gNMIServiceProvider provider{repo, address, port, "admin", "admin"};
    gNMIService gs{};

    build_int_config(provider);

    // Build subscription
    auto i = make_shared<openconfig_interfaces::Interfaces::Interface>();
    i->yfilter = YFilter::read;

    gNMISubscription subscription{};
    subscription.entity = i.get();
    //subscription.subscription_mode = "ON_CHANGE";
    subscription.sample_interval = 10000000000;
    //subscription.suppress_redundant = true;
    subscription.heartbeat_interval = 100000000000;

    gs.subscribe(provider, subscription, 10, "ONCE", "JSON_IETF", gnmi_service_subscribe_callback);
}

TEST_CASE("gnmi_service_subscribe_multiples")
{
    // session
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;

    gNMIServiceProvider provider{repo, address, port, "admin", "admin"};
    gNMIService gs{};

    // Build configuration on the server
    build_int_config(provider);
    build_bgp_config(provider);

    // Build subscription
    auto ifc = make_shared<openconfig_interfaces::Interfaces::Interface>();
    ifc->name = "*";

    auto neighbor = make_shared<openconfig_bgp::Bgp::Neighbors::Neighbor>();
    neighbor->neighbor_address = "*";

    // Build subscriptions
    gNMISubscription ifc_subscription{};
    ifc_subscription.entity = ifc.get();
    //ifc_subscription.subscription_mode = "ON_CHANGE";
    ifc_subscription.sample_interval = 10000000000;
    //ifc_subscription.suppress_redundant = true;
    ifc_subscription.heartbeat_interval = 100000000000;

    gNMISubscription bgp_subscription{};
    bgp_subscription.entity = neighbor.get();
    bgp_subscription.subscription_mode = "TARGET_DEFINED";
    bgp_subscription.sample_interval = 20000000000;
    bgp_subscription.suppress_redundant = false;
    bgp_subscription.heartbeat_interval = 200000000000;

    vector<gNMISubscription*> subscription_list{};
    subscription_list.push_back(&ifc_subscription);
    subscription_list.push_back(&bgp_subscription);

    // Subscribe
    gs.subscribe(provider, subscription_list, 10, "ONCE", "JSON_IETF", gnmi_service_subscribe_multiples_callback);
}

static int counter;

void set_counter(int max)
{
	counter = max;
}

bool counter_poll_request(const char * subscribe_response)
{
	std::this_thread::sleep_for(std::chrono::seconds(2));
	return --counter >= 0;
}

TEST_CASE("gnmi_service_poll_subscribe")
{
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;

    gNMIServiceProvider provider{repo, address, port, "admin", "admin"};
    gNMIService gs{};

    build_int_config(provider);

    // Build subscription
    openconfig_interfaces::Interfaces filter = {};
    auto i = make_shared<openconfig_interfaces::Interfaces::Interface>();
    i->name = "*";
    filter.interface.append(i);

    gNMISubscription subscription{};
    subscription.entity = &filter;

    set_counter(2);
    gs.subscribe(provider, subscription, 10, "POLL", "JSON_IETF", gnmi_service_subscribe_callback, counter_poll_request);
}

TEST_CASE("gnmi_service_stream_subscribe")
{
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;

    gNMIServiceProvider provider{repo, address, port, "admin", "admin"};
    gNMIService gs{};

    build_int_config(provider);

    // Build subscription
    openconfig_interfaces::Interfaces filter = {};
    auto i = make_shared<openconfig_interfaces::Interfaces::Interface>();
    i->name = "*";
    filter.interface.append(i);

    gNMISubscription subscription{};
    subscription.entity = &filter;
    subscription.sample_interval = 2000000000;
    subscription.heartbeat_interval = 10000000000;

    gs.subscribe(provider, subscription, 10, "STREAM", "JSON_IETF", gnmi_service_subscribe_callback);
}

TEST_CASE("gnmi_service_create")
{
    // session
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;

    gNMIServiceProvider provider{repo, address, port, "admin", "admin"};
    gNMIService gs{};
    CodecServiceProvider codec_provider{EncodingFormat::JSON};
    CodecService codec_service{};

    auto ifc = make_shared<openconfig_interfaces::Interfaces::Interface>();
    ifc->name = "Loopback10";
    ifc->config->name = "Loopback10";
    ifc->config->description = "Test";

    openconfig_interfaces::Interfaces ifcs{};
    ifcs.interface.append(ifc);
    ifcs.yfilter = YFilter::replace;

    // Set-replace Request
    auto set_reply = gs.set(provider, ifcs);

    openconfig_interfaces::Interfaces filter{};
    auto get_reply = gs.get(provider, filter, "CONFIG");
    REQUIRE(get_reply != nullptr);
    //print_entity(get_reply, provider.get_session().get_root_schema());
    string expected = R"( <interfaces>
   <interface>
     <name>Loopback10</name>
     <config>
       <name>Loopback10</name>
       <description>Test</description>
     </config>
   </interface>
 </interfaces>
)";
    REQUIRE(entity2string(get_reply, provider.get_session().get_root_schema()) == expected);
}

TEST_CASE("gnmi_service_get_multiple")
{
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;
    gNMIServiceProvider provider{repo, address, port, "admin", "admin"};
    gNMIService gs{};

    // Set Create Request
    openconfig_bgp::Bgp bgp{};
    bgp.yfilter = YFilter::replace;
    config_bgp(bgp);

    auto ifc = make_shared<openconfig_interfaces::Interfaces::Interface>();
    ifc->name = "Loopback10";
    ifc->config->name = "Loopback10";
    ifc->config->description = "Test";

    openconfig_interfaces::Interfaces ifcs{};
    ifcs.yfilter = YFilter::replace;
    ifcs.interface.append(ifc);

    vector<Entity*> set_entities;
    set_entities.push_back(&bgp);
    set_entities.push_back(&ifcs);

    auto set_reply = gs.set(provider, set_entities);

    // Get Request
    openconfig_bgp::Bgp bgp_filter{};
    openconfig_interfaces::Interfaces int_filter{};
    vector<Entity*> filter;
    filter.push_back(&bgp_filter);
    filter.push_back(&int_filter);

    auto get_reply = gs.get(provider, filter, "CONFIG");
    REQUIRE(get_reply.size() == 2);
}

TEST_CASE("gnmi_service_delete")
{
	// session
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;

    gNMIServiceProvider provider{repo, address, port, "admin", "admin"};
    gNMIService gs{};

    openconfig_bgp::Bgp filter = {};
    openconfig_bgp::Bgp bgp = {};

    // Set Delete Request
    bgp.yfilter = YFilter::delete_;
    auto delete_reply = gs.set(provider, bgp);
    REQUIRE(delete_reply);

    // Get Request
    auto get_after_delete_reply = gs.get(provider, filter, "STATE");
    REQUIRE(get_after_delete_reply);
}

TEST_CASE("gnmi_service_get_list_element")
{
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;
    string username = "admin"; string password = "admin";

    gNMIServiceProvider provider{repo, address, port, username, password};
    gNMIService gs{};

    build_int_config(provider);

    auto ifc = make_shared<openconfig_interfaces::Interfaces::Interface>();
    ifc->name = "Loopback10";

    openconfig_interfaces::Interfaces ifcs{};
    ifcs.interface.append(ifc);

    auto get_reply = gs.get(provider, ifcs, "CONFIG");
    REQUIRE(get_reply != nullptr);
    //print_entity(get_reply, provider.get_session().get_root_schema());
    string expected = R"( <interfaces>
   <interface>
     <name>Loopback10</name>
     <config>
       <name>Loopback10</name>
       <description>Test</description>
     </config>
   </interface>
 </interfaces>
)";
    REQUIRE(entity2string(get_reply, provider.get_session().get_root_schema()) == expected);
}

TEST_CASE("gnmi_service_get_leaf")
{
    path::Repository repo{TEST_HOME};
    string address = "127.0.0.1"; int port = 50051;
    string username = "admin"; string password = "admin";

    gNMIServiceProvider provider{repo, address, port, username, password};
    gNMIService gs{};

    build_int_config(provider);

    auto ifc = make_shared<openconfig_interfaces::Interfaces::Interface>();
    ifc->name = "Loopback10";
    ifc->config->description.yfilter = YFilter::read;

    openconfig_interfaces::Interfaces ifcs{};
    ifcs.interface.append(ifc);

    auto get_reply = gs.get(provider, ifcs, "CONFIG");
    REQUIRE(get_reply != nullptr);
    string expected = R"( <interfaces>
   <interface>
     <name>Loopback10</name>
     <config>
       <description>Test</description>
     </config>
   </interface>
 </interfaces>
)";
    REQUIRE(entity2string(get_reply, provider.get_session().get_root_schema()) == expected);
}

File: sdk/python/gnmi/tests/test_gnmi_service.py
#  ----------------------------------------------------------------
# Copyright 2018 Cisco Systems
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------

"""test_gnmi_service.py
Sanity tests for gNMIService
"""

from __future__ import absolute_import
import unittest
import logging

from multiprocessing import Pool

from test_utils import enable_logging, get_local_repo_dir, print_entity, entity_to_string, EmptyTest

from ydk.filters import YFilter
from ydk.path  import Repository
from ydk.types import EncodingFormat, Filter, Config
from ydk.services  import CodecService
from ydk.providers import CodecServiceProvider

from ydk.gnmi.providers import gNMIServiceProvider
from ydk.gnmi.services import gNMIService, gNMISubscription

from ydk.models.ydktest import openconfig_bgp, openconfig_interfaces

int_update = '''json_ietf_val: "{\\"interface\\":[{\\"name\\":\\"Loopback10\\",\\"config\\":{\\"name\\":\\"Loopback10\\",\\"description\\":\\"Test\\"}}]}"'''

bgp_update = '''json_ietf_val: "{\\"global\\":{\\"config\\":{\\"as\\":65172} },\\"neighbors\\":{\\"neighbor\\":[{\\"neighbor-address\\":\\"172.16.255.2\\",\\"config\\":{\\"neighbor-address\\":\\"172.16.255.2\\",\\"peer-as\\":65172}}]}}"'''

def build_bgp_config():
    #Build BGP configuration on server
    bgp = openconfig_bgp.Bgp()
    bgp.global_.config.as_ = 65172
    neighbor = bgp.Neighbors.Neighbor()
    neighbor.neighbor_address = '172.16.255.2'
    neighbor.config.neighbor_address = '172.16.255.2'
    neighbor.config.peer_as = 65172
    bgp.neighbors.neighbor.append(neighbor)
    return bgp

def build_int_config():
    lo10 = openconfig_interfaces.Interfaces.Interface()
    lo10.name = 'Loopback10'
    lo10.config.name = 'Loopback10'
    lo10.config.description = 'Test'
    return lo10

class SanityGnmiService(unittest.TestCase):

    @classmethod
    def setUpClass(self):
        self.codec_provider = CodecServiceProvider()
        self.codec = CodecService()
        self.repo = Repository(get_local_repo_dir())
        self.provider = gNMIServiceProvider( self.repo, "127.0.0.1", 50051, "admin", "admin")
        self.schema = self.provider.get_session().get_root_schema()
        self.gs = gNMIService()

    def test_gnmi_service_capabilities(self):
        json_caps = self.gs.capabilities(self.provider);
        cap_update = '''
      {
        "name": "openconfig-bgp",
        "organization": "OpenConfig working group",
        "version": "2016-06-21"
      },'''
        self.assertEqual((cap_update in json_caps), True)
 
    def test_gnmi_service_set_get(self):
        # Create interface configuration
        ifc = build_int_config()
        ifc.yfilter = YFilter.replace
        reply = self.gs.set(self.provider, ifc)
 
        # Update interface configuration
        ifc.config.description = 'Test'
        ifc.yfilter = YFilter.update
        reply = self.gs.set(self.provider, ifc)
 
        # Get interface configuration
        ifc_filter = openconfig_interfaces.Interfaces.Interface()
        ifc_filter.name = 'Loopback10'
        response = self.gs.get(self.provider, ifc_filter, "CONFIG")
        self.assertIsNotNone(response)
        self.assertTrue(hasattr(response, 'name'))
        self.assertEqual(response.name, 'Loopback10')
 
        # Get interface description only
        ifc_filter = openconfig_interfaces.Interfaces.Interface()
        ifc_filter.name = 'Loopback10'
        ifc_filter.config.description = YFilter.read

        response = self.gs.get(self.provider, ifc_filter, "CONFIG")
        self.assertIsNotNone(response)
        #print_entity(ifc_config, self.schema)
        expected = '''<interface>
  <name>Loopback10</name>
  <config>
    <name>Loopback10</name>
    <description>Test</description>
  </config>
</interface>
'''
        self.assertEqual( entity_to_string(response, self.schema), expected)
 
        # Delete interface configuration
        ifc_delete = openconfig_interfaces.Interfaces.Interface()
        ifc_delete.name = 'Loopback10'
        ifc_delete.yfilter = YFilter.delete
        reply = self.gs.set(self.provider, ifc_delete)
 
    def test_gnmi_service_set_get_multiple(self):
        # Create interface and BGP configuration
        ifc = build_int_config()
        ifc.yfilter = YFilter.replace
        bgp = build_bgp_config()
        bgp.yfilter = YFilter.replace
        reply = self.gs.set(self.provider, Config(ifc, bgp))
 
        # Get and print interface and BGP configuration
        ifc_filter = openconfig_interfaces.Interfaces()
        bgp_filter = openconfig_bgp.Bgp()
        response = self.gs.get(self.provider, Filter(ifc_filter, bgp_filter), "CONFIG")
        self.assertIsNotNone(response)
        self.assertEqual(response.__len__(), 2)
        #for entity in response:
        #    print_entity(entity, self.schema)
 
        # Delete interface and BGP configuration
        ifc.yfilter = YFilter.delete
        bgp.yfilter = YFilter.delete
        reply = self.gs.set(self.provider, [ifc, bgp])
  
    def test_gnmi_service_subscribe_once(self):
        # Create BGP configuration
        bgp = build_bgp_config()
        bgp.yfilter = YFilter.replace
        reply = self.gs.set(self.provider, bgp)
 
        subscription = gNMISubscription()
        bgp_filter = openconfig_bgp.Bgp()
        subscription.subscription_mode = "ON_CHANGE"
        subscription.sample_interval = 10 * 1000000000
        subscription.suppress_redundant = True
        subscription.heartbeat_interval = 100 * 1000000000
        subscription.entity = bgp_filter
 
        self.gs.subscribe(self.provider, subscription, 10, "ONCE", "JSON_IETF", gnmi_subscribe_callback);
 
        # Delete BGP configuration
        bgp.yfilter = YFilter.delete
        reply = self.gs.set(self.provider, bgp)
 
    def test_gnmi_service_subscribe_multiple(self):
        # Create interface and BGP configurations
        bgp = build_bgp_config()
        bgp.yfilter = YFilter.replace
 
        ifc = build_int_config()
        ifc.yfilter = YFilter.replace
        reply = self.gs.set(self.provider, [bgp, ifc])
         
        bgp_filter = openconfig_bgp.Bgp()
        bgp_subscription = gNMISubscription()
        bgp_subscription.entity = bgp_filter
 
        int_filter = openconfig_interfaces.Interfaces()
        int_subscription = gNMISubscription()
        int_subscription.entity = int_filter
 
        self.gs.subscribe(self.provider, [int_subscription, bgp_subscription], 10, "ONCE", "JSON_IETF", gnmi_subscribe_multiples_callback)
 
        # Delete BGP configuration
        bgp.yfilter = YFilter.delete
        ifc.yfilter = YFilter.delete
        reply = self.gs.set(self.provider, [bgp, ifc])
 
    def test_gnmi_service_subscribe_stream(self):
        # Create BGP configuration
        bgp = build_bgp_config()
        bgp.yfilter = YFilter.replace
        reply = self.gs.set(self.provider, bgp)

        subscription = gNMISubscription()
        bgp_filter = openconfig_bgp.Bgp()
        subscription.subscription_mode = "SAMPLE"
        subscription.sample_interval = 2 * 1000000000
        subscription.suppress_redundant = False
        subscription.heartbeat_interval = 8 * 1000000000
        subscription.entity = bgp_filter

        self.gs.subscribe(self.provider, subscription, 10, "STREAM", "JSON_IETF", gnmi_subscribe_callback)

        # Delete BGP configuration
        bgp.yfilter = YFilter.delete
        reply = self.gs.set(self.provider, bgp)

    def test_gnmi_service_subscribe_poll(self):
        # Create BGP configuration
        bgp = build_bgp_config()
        bgp.yfilter = YFilter.replace
        reply = self.gs.set(self.provider, bgp)

        subscription = gNMISubscription()
        bgp_filter = openconfig_bgp.Bgp()
        subscription.entity = bgp_filter

        self.gs.subscribe(self.provider, subscription, 10, "POLL", "JSON_IETF", gnmi_subscribe_callback)

        # Delete BGP configuration
        bgp.yfilter = YFilter.delete
        reply = self.gs.set(self.provider, bgp)

def read_sub(response):
    print("===> In response callback. Received subscribe response:")
    print(response)

def gnmi_subscribe_callback(response):
    #read_sub(response)
    test = EmptyTest()
    test.assertTrue(bgp_update in response)

def gnmi_subscribe_multiples_callback(response):
    test = EmptyTest()
    test.assertTrue(int_update in response)
    test.assertTrue(bgp_update in response)

if __name__ == '__main__':
    import sys
    enable_logging(logging.ERROR)
    suite = unittest.TestLoader().loadTestsFromTestCase(SanityGnmiService)
    ret = not unittest.TextTestRunner(verbosity=2).run(suite).wasSuccessful()
    sys.exit(ret)
